### PR TITLE
switch to electron-builder to code sign and notarize Cypress application

### LINKED ISSUE #5791: “Cypress” can’t be opened because Apple cannot check it for malicious software.
I try to open this after download and I see this message on mac.
------------------------------------------------------
“Cypress” can’t be opened because Apple cannot check it for malicious software.

This software needs to be updated. Contact the developer for more information.

Firefox downloaded this file today at 1:28 PM from docs.cypress.io.

------------------------------------------------------
### Current behavior:

“Cypress” can’t be opened because Apple cannot check it for malicious software.

This software needs to be updated. Contact the developer for more information.

Firefox downloaded this file today at 1:28 PM from docs.cypress.io.

### Desired behavior:

able to open and use after download 

### Steps to reproduce: (app code and test code)

1. download from here 
https://download.cypress.io/desktop

2. open after download

### Versions

firefox 68.2.0esr(64-bit)
mac os catalina 10.15

### PR DESCRIPTION
**Pull request by bahmutov**
- Closes #5791

### User facing changelog

- when running Cypress on Mac OS, there should be no warnings from Mac security tools
- Mac application is signed using the new Cypress.io application certificate
- Update `execa` from `1.0.0` to `4.0.0`

### Additional details

Mac OS Catalina requires the apps to be distributed to use new hardened runtime and be notarized. We also need to switch to the new company application certificate to code sign the app, instead of a personal certificate that is close to expiring

### How has the user experience changed?

No changes

### PR Tasks

- [ ] Has the original issue been tagged with a release in ZenHub? 

## Details

- [x] builds on Mac CI
  - [x] binary verify step does not break notarization #6089
- [x] builds on Linux CI
- [x] builds on Windows CI

### stretch / unimportant

- [x] build on Mac does not create extra files / zip archives
- [x] Linux build does not create extra files / zip archives

### CODE REVIEW

**Comment by bahmutov** [comment]
Investigating zip size increase on Linux, seems I am not deleting or including Electron app several times

**Comment by bahmutov** [comment]
Current branch has extra `dist folder
```
root@4502d25b8c39:/home/person/cypress/build# du -hs linux-unpacked/resources/app/packages/electron/*
292M	linux-unpacked/resources/app/packages/electron/dist
4.0K	linux-unpacked/resources/app/packages/electron/index.js
12K	linux-unpacked/resources/app/packages/electron/lib
28M	linux-unpacked/resources/app/packages/electron/node_modules
4.0K	linux-unpacked/resources/app/packages/electron/package.json
```
which is not in the `develop` branch
```
root@f7a93d5c4579:/home/person/cypress# du -hs build/linux/Cypress/resources/app/packages/electron/*
4.0K	build/linux/Cypress/resources/app/packages/electron/index.js
12K	build/linux/Cypress/resources/app/packages/electron/lib
28M	build/linux/Cypress/resources/app/packages/electron/node_modules
44K	build/linux/Cypress/resources/app/packages/electron/package-lock.json
4.0K	build/linux/Cypress/resources/app/packages/electron/package.json
```
We are probably not unlinking it correctly it anymore

**Comment by bahmutov** [comment]
Comparing installed Linux build right now (3.8.1 is the new version, 3.8.0 is the last release from `develop` branch)

```
root@ff7a4bb48725:~# ls -la /root/.cache/Cypress/3.8.1/linux-unpacked/
total 266896
drwxr-xr-x 5 root root      4096 Dec 20 10:30 .
drwxr-xr-x 3 root root      4096 Dec 20 10:58 ..
-rwxr-xr-x 1 root root 115375408 Dec 20 10:30 Cypress
-rw-r--r-- 1 root root      1060 Dec 20 10:30 LICENSE.electron.txt
-rw-r--r-- 1 root root   4675743 Dec 20 10:30 LICENSES.chromium.html
-rwxr-xr-x 1 root root   5581136 Dec 20 10:30 chrome-sandbox
-rw-r--r-- 1 root root    178136 Dec 20 10:30 chrome_100_percent.pak
-rw-r--r-- 1 root root    316984 Dec 20 10:30 chrome_200_percent.pak
-rwxr-xr-x 1 root root 115375408 Dec 20 10:30 cypress
-rw-r--r-- 1 root root  10426784 Dec 20 10:30 icudtl.dat
-rwxr-xr-x 1 root root    248168 Dec 20 10:30 libEGL.so
-rwxr-xr-x 1 root root   8375488 Dec 20 10:30 libGLESv2.so
-rwxr-xr-x 1 root root   2973944 Dec 20 10:30 libffmpeg.so
drwxr-xr-x 2 root root      4096 Dec 20 10:30 locales
-rw-r--r-- 1 root root     82118 Dec 20 10:30 natives_blob.bin
drwxr-xr-x 3 root root      4096 Dec 20 10:30 resources
-rw-r--r-- 1 root root   8646888 Dec 20 10:30 resources.pak
-rw-r--r-- 1 root root    280392 Dec 20 10:30 snapshot_blob.bin
drwxr-xr-x 2 root root      4096 Dec 20 10:30 swiftshader
-rw-r--r-- 1 root root    701056 Dec 20 10:30 v8_context_snapshot.bin
root@ff7a4bb48725:~# ls -la /root/.cache/Cypress/3.8.0/Cypress/       
total 154220
drwxr-xr-x 5 root root      4096 Dec 12 17:09 .
drwxr-xr-x 3 root root      4096 Dec 20 11:00 ..
-rwxr-xr-x 1 root root 115375408 Dec  9 22:48 Cypress
-rw-r--r-- 1 root root      1060 Dec  9 21:35 LICENSE
-rw-r--r-- 1 root root   4675743 Dec  9 22:11 LICENSES.chromium.html
-rwxr-xr-x 1 root root   5581136 Dec  9 22:48 chrome-sandbox
-rw-r--r-- 1 root root    178136 Dec  9 22:26 chrome_100_percent.pak
-rw-r--r-- 1 root root    316984 Dec  9 22:26 chrome_200_percent.pak
-rw-r--r-- 1 root root  10426784 Dec  9 22:06 icudtl.dat
-rwxr-xr-x 1 root root    248168 Dec  9 22:48 libEGL.so
-rwxr-xr-x 1 root root   8375488 Dec  9 22:48 libGLESv2.so
-rwxr-xr-x 1 root root   2973944 Dec  9 22:48 libffmpeg.so
drwxr-xr-x 2 root root      4096 Dec 12 17:09 locales
-rw-r--r-- 1 root root     82118 Dec  9 22:35 natives_blob.bin
drwxr-xr-x 3 root root      4096 Dec 12 17:09 resources
-rw-r--r-- 1 root root   8646888 Dec  9 22:32 resources.pak
-rw-r--r-- 1 root root    280392 Dec  9 22:45 snapshot_blob.bin
drwxr-xr-x 2 root root      4096 Dec 12 17:09 swiftshader
-rw-r--r-- 1 root root    701056 Dec  9 22:45 v8_context_snapshot.bin
-rw-r--r-- 1 root root         5 Dec  9 21:35 version
```
Two things are different
1. the folder is named `linux-unpacked` and should be called `Cypress`
2. inside the folder `linux-unpacked` you have two identical files - one is `Cypress` another `cypress`. It should only have `Cypress` file - which is the built Electron binary

**Comment by bahmutov** [comment]
Trying to build on Windows machine
```
npm run all build-js --skip-packages cli

> cypress@3.8.0 all c:\git\cypress
> node ./scripts/run.js "build-js" "cli"

filtered packages: c:\git\cypress\packages\{launcher, network, proxy, server}
[launcher:build-js]
[launcher:build-js] > @packages/launcher@0.0.0 build-js c:\git\cypress\packages\launcher
[launcher:build-js] > bin-up tsc --project .
[launcher:build-js]
[network:build-js ]
[network:build-js ] > @packages/network@0.0.0 build-js c:\git\cypress\packages\network
[network:build-js ] > bin-up tsc --project .
[network:build-js ]
[server:build-js  ]
[server:build-js  ] > @packages/server@0.0.0 build-js c:\git\cypress\packages\server
[server:build-js  ] > tsc
[server:build-js  ]
[launcher:build-js] Found c:\git\cypress\node_modules\.bin\tsc
[network:build-js ] Found c:\git\cypress\node_modules\.bin\tsc
[proxy:build-js   ]
[proxy:build-js   ] > @packages/proxy@0.0.0 build-js c:\git\cypress\packages\proxy
[proxy:build-js   ] > bin-up tsc --project .
[proxy:build-js   ]
[proxy:build-js   ] Found c:\git\cypress\node_modules\.bin\tsc
[server:build-js  ] lib/util/stream_buffer.ts(62,13): error TS2339: Property 'createReadStream' does not exist on type 'Writable'.
[server:build-js  ] lib/util/stream_buffer.ts(127,13): error TS2339: Property 'unpipeAll' does not exist on type 'Writable'.
[server:build-js  ] lib/util/stream_buffer.ts(132,13): error TS2339: Property '_buffer' does not exist on type 'Writable'.
[server:build-js  ] lib/util/stream_buffer.ts(136,13): error TS2339: Property '_finished' does not exist on type 'Writable'.
[server:build-js  ] npm ERR! code ELIFECYCLE
[server:build-js  ] npm ERR! errno 2
[server:build-js  ] npm ERR! @packages/server@0.0.0 build-js: `tsc`
[server:build-js  ] npm ERR! Exit status 2
[server:build-js  ] npm ERR!
[server:build-js  ] npm ERR! Failed at the @packages/server@0.0.0 build-js script.
[server:build-js  ] npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
[server:build-js  ]
[server:build-js  ] npm ERR! A complete log of this run can be found in:
[server:build-js  ] npm ERR!     C:\Users\Administrator\AppData\Roaming\npm-cache\_logs\2019-12-27T20_10_11_889Z-debug.log

One or more tasks failed running 'npm run all build-js'.

Results:

.----------------------.
| package  | exit code |
|----------|-----------|
| launcher |           |
| network  |           |
| proxy    |           |
| server   |         2 |
'----------------------'

stderr:
```

**Comment by bahmutov** [comment]
ughh, what the ef, the Windows was fine, but on Appveyor x86
```
by calling: node index.js --version
in the folder C:\projects\cypress\build\win-unpacked\resources\app
🔥 deploy error
Error: spawn node ENOENT
    at Process.ChildProcess._handle.onexit (internal/child_process.js:248:19)
    at onErrorNT (internal/child_process.js:431:16)
    at processTicksAndRejections (internal/process/task_queues.js:83:17)
 {
  errno: 'ENOENT',
  code: 'ENOENT',
  syscall: 'spawn node',
  path: 'node',
  spawnargs: [ 'index.js', '--version' ],
  stdout: '',
  stderr: '',
  failed: true,
  signal: null,
  cmd: 'node index.js --version',
  timedOut: false,
  killed: false
}
npm ERR! code ELIFECYCLE
npm ERR! errno 1
```

**Comment by bahmutov** [comment]
WHY WHY WHY is it again not notarized!!!!

Screen Shot 2020-01-03 at 2 43 30 PM

Build https://circleci.com/gh/cypress-io/cypress/226113 shows

```
  electron-osx-sign Application signed. +0ms
afterSign hook triggered in /Users/distiller/cypress/build/mac
Notarizing com.electron.cypress found at /Users/distiller/cypress/build/mac/Cypress.app
Done notarizing com.electron.cypress

1:25:23 PM #verifyAppCanOpen /Users/distiller/cypress/build/mac/Cypress.app darwin
/Users/distiller/cypress/build/mac/Cypress.app: accepted
source=Developer ID
origin=Developer ID Application: Cypress.Io, Inc. (7D655LWGLY)
```

**Comment by bahmutov** [comment]
You can test the new Mac binary locally by installing it like this (delete previous install of this version if any):
```
CYPRESS_INSTALL_BINARY=https://cdn.cypress.io/beta/binary/3.8.2/darwin-x64/circle-try-new-electron-sign-37b5625ff3ccc82c2341be5f3e11e8bff2a82def-228042/cypress.zip npm install https://cdn.cypress.io/beta/npm/3.8.2/circle-try-new-electron-sign-37b5625ff3ccc82c2341be5f3e11e8bff2a82def-227975/cypress.tgz
```

**Comment by bahmutov** [comment]
Hmm, looks like verify does not save or find the right binary state file
Screen Shot 2020-01-08 at 12 45 58 PM

File: circle.yml [hunk 1]
```diff
@@ -8,6 +8,7 @@ macBuildFilters: &macBuildFilters
     branches:
       only:
         - develop
+        - try-new-electron-sign
```
Comment (Bkucera) [inline circle.yml]: is this intended to be removed before merge?

**Comment by bahmutov** [comment]
Sure I can remove it before merging. Or it can be removed afterwards

File: scripts/after-sign-hook.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+  try {
+    await electron_notarize.notarize({
+      appBundleId: appId,
+      appPath,
+      appleId: process.env.NOTARIZE_APP_APPLE_ID,
+      appleIdPassword: process.env.NOTARIZE_APP_PASSWORD,
+    })
+  } catch (error) {
+    console.error(error)
```
Comment (brian-mann) [inline scripts/after-sign-hook.js]: shouldn't errors thrown by the notarization tool crash the process so CI exits correctly?

as it stands... wouldn't this silently pass in CI and probably cause all kinds of problems downstream?

File: scripts/binary/build.coffee [hunk 3]
```diff
@@ -229,18 +296,84 @@ buildCypressApp = (platform, version, options = {}) ->
...
+    fs.removeAsync(electronDistFolder) # .catch(_.noop) why are we ignoring an error here?!
+
+  removeDuplicateElectron = ->
+    log('#removeDuplicateElectron')
+    if platform != "linux"
+      console.log("no duplicate file on platform #{platform}")
+      return Promise.resolve()
+    # for some reason I see duplicate "electron" binary left in the build folder
+    # build/linux-unpacked/Cypress <-- good built app
+    # build/linux-unpacked/electron <-- same file as "Cypress" app
```
Comment (brian-mann) [inline scripts/binary/build.coffee]: this sounds really weird...

File: scripts/binary/build.coffee [hunk 4]
```diff
@@ -229,18 +296,84 @@ buildCypressApp = (platform, version, options = {}) ->
...
+        fs.promises.stat(cypressFilename),
+      ])
+    .then ([CypressStats, cypressStats]) ->
+      console.log('Cypress stats for', CypressFilename)
+      console.log("%o", CypressStats)
+
+      console.log('cypress stats for', cypressFilename)
+      console.log("%o", CypressStats)
+
+      if CypressStats.ino == cypressStats.ino
```
Comment (brian-mann) [inline scripts/binary/build.coffee]: whys this necessary? don't we always want to remove the duplicate?

File: scripts/binary/build.coffee [hunk 5]
```diff
@@ -229,18 +296,84 @@ buildCypressApp = (platform, version, options = {}) ->
...
+  removeDuplicateElectron = ->
+    log('#removeDuplicateElectron')
+    if platform != "linux"
+      console.log("no duplicate file on platform #{platform}")
+      return Promise.resolve()
+    # for some reason I see duplicate "electron" binary left in the build folder
+    # build/linux-unpacked/Cypress <-- good built app
+    # build/linux-unpacked/electron <-- same file as "Cypress" app
+    CypressFilename = buildDir("Cypress")
+    cypressFilename = buildDir("cypress")
```
Comment (brian-mann) [inline scripts/binary/build.coffee]: you mention previously that `Cypress` and `electron` are the duplicates... but your code is checking for `Cypress` and `cypress`

File: scripts/binary/zip.coffee [hunk 6]
```diff
@@ -8,29 +8,38 @@ fs      = require("fs")
...
+      # --sequesterRsrc When creating a PKZip archive, preserve resource
+      #   forks and HFS meta-data in the subdirectory __MACOSX
+      # --keepParent when zipping folder "foo", makes the folder
+      #   the top level in the archive
+      #   foo.zip
+      #     foo/
+      #        ...
+      zip = "ditto -c -k --sequesterRsrc --keepParent #{src} #{dest}"
+      console.log(zip)
+      cp.exec zip, {}, (err, stdout, stderr) ->
```
Comment (brian-mann) [inline scripts/binary/zip.coffee]: switch to using `execa` here?

**Review by brian-mann — CHANGES_REQUESTED** [review]

**Comment by brian-mann** [comment]
@bahmutov if you can get an update on this and address the feedback we can get this merged in

**Comment by brian-mann** [comment]
@bahmutov can i get an update on this? would like to have this PR merged in by the next release

**Comment by bahmutov** [comment]
I did not have time to do this, I suggest this goes after 4.0

**Comment by bahmutov** [comment]
Path `--c.electronDist=/Users/gleb/git/cypress/packages/electron/node_modules/electron/dist` seems incorrect, does not exist

**Comment by bahmutov** [comment]
tried the latest notarization, getting the following error on Mac

Screen Shot 2020-03-02 at 12 49 14 PM

**Comment by bahmutov** [comment]
Upgraded `electron-builder` to the latest version, still the same problems.

1. timestamp

```
{
"severity": "error",
"code": null,
"path": "Cypress.zip/Cypress.app/Contents/Resources/app/packages/server/node_modules/babel-plugin-add-module-exports/node_modules/fsevents/build/Release/fse.node",
"message": "The signature does not include a secure timestamp.",
"docUrl": null,
"architecture": "x86_64"
},
```

The page https://developer.apple.com/documentation/xcode/notarizing_macos_software_before_distribution/resolving_common_notarization_issues suggests passing `--timestamp` flag when codesigning, trying to figure out how to do this.

**Comment by bahmutov** [comment]
Just tried a previous commit https://github.com/cypress-io/cypress/pull/6013/commits/37b5625ff3ccc82c2341be5f3e11e8bff2a82def that used to work - and got the same signing problems https://osxapps-ssl.itunes.apple.com/itunes-assets/Enigma114/v4/fd/d3/c8/fdd3c892-982b-0b9e-c8a9-c88d0186e8d8/developer_log.json?accessKey=1583457539_1501347668417433065_w3ydcWK%2F5dUfX5g2j3xIEGLGZxT5E2se0xqQcsrSCGQwWGwkJRRycucZdPRjX%2Bw7r%2FxucajhOoEWsmMNE%2FqKrPetrgibFj4y%2BpTvfroBwktR8uL%2FTGt%2BVwEv7hYN7vjScOCXY6TrWEEpUYy1T5QIgxxGrxilVKPWitTatSAHfvQ%3D

So it seems this is something Apple has added to their notarization checks to stop binary files from being included

**Comment by bahmutov** [comment]
Current errors all seem to be because the archive includes binaries:

Cypress.zip/Cypress.app/Contents/Resources/app/packages/server/node_modules/@ffmpeg-installer/darwin-x64/ffmpeg
- `The binary is not signed.`
- `The signature does not include a secure timestamp.`
- `The executable does not have the hardened runtime enabled.`

Then the next file Cypress.zip/Cypress.app/Contents/Resources/app/packages/server/node_modules/registry-js/build/Release/registry.node
- `The binary is not signed.`
- `The signature does not include a secure timestamp.`

Cypress.zip/Cypress.app/Contents/Resources/app/packages/server/node_modules/chokidar/node_modules/fsevents/fsevents.node
- `The binary is not signed.`
- `The signature does not include a secure timestamp.`

then similar errors for Cypress.zip/Cypress.app/Contents/Resources/app/packages/server/node_modules/term-size/vendor/macos/term-size and Cypress.zip/Cypress.app/Contents/Resources/app/packages/server/node_modules/trash/lib/macos-trash and Cypress.zip/Cypress.app/Contents/Resources/app/packages/server/node_modules/fsevents/build/Release/.node and Cypress.zip/Cypress.app/Contents/Resources/app/packages/server/node_modules/fsevents/build/Release/fse.node

Idea: can we add these binaries to the list of binaries signed by the codesign?!

**Comment by bahmutov** [comment]
I have added the additional binaries to sign like ffmpeg to the electron-builder.json, unfortunately, could not find a way to use wildcards, so had to list each file separately

**Comment by bahmutov** [comment]
Packing for Linux has changed the output (build) folder names

Current version in `develop`

```
root@3b662bb808a6:~/cypress# ls -la build/linux/ 
total 12
drwxr-xr-x 3 root root 4096 Mar  5 12:01 .
drwxr-xr-x 3 root root 4096 Mar  5 12:01 ..
drwxr-xr-x 5 root root 4096 Mar  5 12:01 Cypress

root@3b662bb808a6:~/cypress# ls -la build/linux/Cypress/
total 154296
drwxr-xr-x 5 root root      4096 Mar  5 12:01 .
drwxr-xr-x 3 root root      4096 Mar  5 12:01 ..
-rwxr-xr-x 1 root root 115449136 Feb 20 18:53 Cypress
-rw-r--r-- 1 root root      1060 Feb 20 17:38 LICENSE
-rw-r--r-- 1 root root   4675743 Feb 20 18:15 LICENSES.chromium.html
-rwxr-xr-x 1 root root   5581136 Feb 20 18:53 chrome-sandbox
-rw-r--r-- 1 root root    178136 Feb 20 18:33 chrome_100_percent.pak
-rw-r--r-- 1 root root    316984 Feb 20 18:33 chrome_200_percent.pak
-rw-r--r-- 1 root root  10426784 Feb 20 18:09 icudtl.dat
-rwxr-xr-x 1 root root    248168 Feb 20 18:53 libEGL.so
-rwxr-xr-x 1 root root   8375488 Feb 20 18:53 libGLESv2.so
-rwxr-xr-x 1 root root   2973944 Feb 20 18:53 libffmpeg.so
drwxr-xr-x 2 root root      4096 Mar  5 12:01 locales
-rw-r--r-- 1 root root     82118 Feb 20 18:40 natives_blob.bin
drwxr-xr-x 3 root root      4096 Mar  5 12:01 resources
-rw-r--r-- 1 root root   8646888 Feb 20 18:39 resources.pak
-rw-r--r-- 1 root root    280392 Feb 20 18:50 snapshot_blob.bin
drwxr-xr-x 2 root root      4096 Mar  5 12:01 swiftshader
-rw-r--r-- 1 root root    701056 Feb 20 18:50 v8_context_snapshot.bin
-rw-r--r-- 1 root root         6 Feb 20 17:38 version
```

This PR has instead

```
root@8033c5848e49:~/cypress# ls -la build/linux-unpacked/
total 154288
drwxr-xr-x 5 root root      4096 Mar  5 12:00 .
drwxr-xr-x 3 root root      4096 Mar  5 12:00 ..
-rw-r--r-- 1 root root      1060 Mar  5 12:00 LICENSE.electron.txt
-rw-r--r-- 1 root root   4675743 Mar  5 12:00 LICENSES.chromium.html
-rwxr-xr-x 1 root root   5581136 Mar  5 12:00 chrome-sandbox
-rw-r--r-- 1 root root    178136 Mar  5 12:00 chrome_100_percent.pak
-rw-r--r-- 1 root root    316984 Mar  5 12:00 chrome_200_percent.pak
-rwxr-xr-x 1 root root 115449136 Mar  5 12:00 cypress
-rw-r--r-- 1 root root  10426784 Mar  5 12:00 icudtl.dat
-rwxr-xr-x 1 root root    248168 Mar  5 12:00 libEGL.so
-rwxr-xr-x 1 root root   8375488 Mar  5 12:00 libGLESv2.so
-rwxr-xr-x 1 root root   2973944 Mar  5 12:00 libffmpeg.so
drwxr-xr-x 2 root root      4096 Mar  5 12:00 locales
-rw-r--r-- 1 root root     82118 Mar  5 12:00 natives_blob.bin
drwxr-xr-x 3 root root      4096 Mar  5 12:00 resources
-rw-r--r-- 1 root root   8646888 Mar  5 12:00 resources.pak
-rw-r--r-- 1 root root    280392 Mar  5 12:00 snapshot_blob.bin
drwxr-xr-x 2 root root      4096 Mar  5 12:00 swiftshader
-rw-r--r-- 1 root root    701056 Mar  5 12:00 v8_context_snapshot.bin
```

**Comment by bahmutov** [comment]
- [ ] blocked: yarn hoists ffmpeg installer out of `packages/server`, which seems  to be a mistake since it is a production dependency

**Comment by bahmutov** [comment]
```
$ ls node_modules/\@ffmpeg-installer/
darwin-x64
~/git/cypress on try-new-electron-sign*
$ ls packages/server/node_modules/\@ffmpeg-installer/
ffmpeg
```

**Comment by bahmutov** [comment]
build and signing is working again, would be nice to merge and get Mac users to install pre-releases this week before using the new code signing certificate

**Review by brian-mann — APPROVED** [review]

**Pull request merged**